MATLAB/Algorithms/LSQR.m

```diff
@@ -0,0 +1,273 @@
+qualMeasOut=zeros(length(QualMeasOpts),niter);
+if redundancy_weights
+    % Data redundancy weighting, W_r implemented using Wang weighting
+    % reference: https://iopscience.iop.org/article/10.1088/1361-6560/ac16bc
+    
+    num_frames = size(proj,3);
+    W_r = redundancy_weighting(geo);
+    W_r = repmat(W_r,[1,1,num_frames]);
+    % disp('Size of redundancy weighting matrix');
+    % disp(size(W_r));
+end
+
+% Paige and Saunders //doi.org/10.1145/355984.355989
+
+% Enumeration as given in the paper for 'Algorithm LSQR'
+% (1) Initialize 
+u=proj-Ax(x,geo,angles,'Siddon','gpuids',gpuids);
+normr = norm(u(:),2);
+u = u/normr;
+
+beta = normr;
+phibar = beta;
+
+% msl: I have not checked 'redundancy_weights', do they have to do with 
+% mismatched transpose? 
+if redundancy_weights 
+    v=Atb(W_r .* u,geo,angles,'matched','gpuids',gpuids);
+else
+    v=Atb(u,geo,angles,'matched','gpuids',gpuids);
+end
+
+alpha = norm(v(:),2); % msl: do we want to check if it is 0? 
+v = v/alpha;
+rhobar = alpha;
+w = v;
+
+normAtr = beta*alpha; % Do we want this? ||A^T r_k||
+% msl: do we want to check for convergence? In well posed problems it would
+% make sense, not sure now.
+
+errorL2=zeros(1,niter); % msl: is this error the residual norm ? 
+
+% (2) Start iterations 
+for ii=1:niter
+    x0 = x;
+    if (ii==1 && verbose);tic;end
+    
+    % (3)(a)
+    u = Ax(v,geo,angles,'Siddon','gpuids',gpuids) - alpha*u;
+    beta = norm(u(:),2);
+    u = u / beta;
+    
+    % (3)(b)
+    if redundancy_weights % msl: Is this right here?
+        v = Atb(W_r .* u,geo,angles,'matched','gpuids',gpuids) - beta*v;
+    else
+        v = Atb(u,geo,angles,'matched','gpuids',gpuids) - beta*v;
+    end
+    alpha = norm(v(:),2);
+    v = v / alpha;    
+
+    % (4)(a-g)
+    rho = sqrt(rhobar^2 + beta^2);
+    c = rhobar / rho;
+    s = beta / rho;
+    theta = s * alpha;
+    rhobar = - c * alpha; % msl: I am overwriting a matlab function, maybe change name
+    phi = c * phibar;
+    phibar = s * phibar;
+    
+    % (5) Update x, w
+    x = x + (phi / rho) * w;
+    w = v - (theta / rho) * w;
+
+    % Update estimated quantities of interest.
+    % msl: We can also compute cheaply estimates of ||x||, ||A||, cond(A)
+    normr = normr*abs(s);   % ||r_k|| = ||b - A x_k||
+                            % Only exact if we do not have orth. loss 
+    normAtr = phibar * alpha * abs(c); % msl: Do we want this? ||A^T r_k||
+    errorL2(ii)=normr;
+    
+    % (6) Test for convergence. 
+    % msl: I still need to implement this. 
+    % msl: There are suggestions on the original paper. Let's talk about it!
+    
+    if measurequality % msl: what is this??
+        qualMeasOut(:,ii)=Measure_Quality(x0,x,QualMeasOpts);
+    end
+
+    if ii>1 && errorL2(ii)>errorL2(ii-1)  % msl: not checked
+        % OUT!
+       x=x-alpha*v;
+       if verbose
+          disp(['CGLS stoped in iteration N', num2str(ii),' due to divergence.']) 
+       end
+       return; 
+    end

```

Comment:
we can remove this maybe, not sure. In CGLS sometimes it would go out, so I stopped. Never sure if it was a math issue or a numerical issue.

Reply:
So this is not a stopping criterion as such but we are just checking the algorithm is not doing weird stuff? (i.e. loss of orthogonality). We should still have some sanity checks maybe (check if the values of the projected matrix are too small?). ErrorL2 is not the error but the residual norm, right? (or the norm of the discrepancy, value of the fit the data term, etc).

Let's try some examples and see what happens, what do you think?

Reply:
Exactly, algorithm not doing weird stuff (I mean, it should never increase the residual, right?). Indeed its the residual norm. 

The problem is that I never seen this with synthetic data, only with real data. In real data the system matrix describes the measurement worse, thus these things happen more.